### PR TITLE
[CALCITE-3779] Implement bitand scalar function

### PR DESCRIPTION
**Pull request by wangxlong**
Implement bitand scalar function which support tinyint, smallint, int, bigint, binary and varbinary type.
Related issue: [CALCITE-3779](https://issues.apache.org/jira/browse/CALCITE-3779)
Parent issue: [CALCITE-3732](https://issues.apache.org/jira/browse/CALCITE-3732)

### CODE REVIEW

**Comment by zinking** [comment]
looks good to me

**Review by hsyuan — COMMENTED** [review]
@wangxlong Can you rebase on latest master?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -1066,12 +1067,53 @@ private static RuntimeException notComparable(String op, Object b0,
...
+  public static ByteString bitAnd(ByteString b0, ByteString b1) {
+
+    if (b0.length() != b1.length()) {
+      throw RESOURCE.invalidInputForBitwiseBinaryValue().ex();
+    }
+    byte[] bytes0 = b0.getBytes();
+    byte[] bytes1 = b1.getBytes();
+    byte[] result = new byte[b0.length()];
+
+    IntStream.range(0, bytes0.length).forEach(i -> {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1087]: Frankly speaking, a good old `for` loop would look better here, and it looks like streams add complexity without much benefits.

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1087]: Can't agree more.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 2]
```diff
@@ -410,6 +411,7 @@ public int size() {
   public static final SqlReturnTypeInference INTEGER_QUOTIENT_NULLABLE =
       chain(ARG0_INTERVAL_NULLABLE, LEAST_RESTRICTIVE);
 
+
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:414]: Please refrain from altering non-related lines

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 3]
```diff
@@ -1066,12 +1067,53 @@ private static RuntimeException notComparable(String op, Object b0,
...
 
+  /** SQL <code>BITAND</code> operator applied to ByteString values */
+  public static ByteString bitAnd(ByteString b0, ByteString b1) {
+
+    if (b0.length() != b1.length()) {
+      throw RESOURCE.invalidInputForBitwiseBinaryValue().ex();
+    }
+    byte[] bytes0 = b0.getBytes();
+    byte[] bytes1 = b1.getBytes();
+    byte[] result = new byte[b0.length()];
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1085]: Technically speaking, `b0.getBytes()` clones the byte array, so there's no need to allocate `new byte[...]` for the result. I believe `bytes0` would work just fine for the resulting array.

**Comment by wangxlong** [comment]
Hi @hsyuan @vlsi I have rebased on latest master.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 4]
```diff
@@ -1120,6 +1120,16 @@ public static ByteString bitAnd(ByteString b0, ByteString b1) {
     return binaryOperator(b0, b1, (x, y) -> (byte) (x & y));
   }
 
+  /** Bitwise function <code>BIT_AND</code> applied to long and binary values. */
+  public static ByteString bitAnd(long b0, ByteString b1) {
+    return binaryOperator(b1, b0, (x, y) -> (byte) (x & y));
+  }
+
+  /** Bitwise function <code>BIT_AND</code> applied to binary and long values. */
+  public static ByteString bitAnd(ByteString b0,  long b1) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1129]: extra space before `long`

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 5]
```diff
@@ -3490,4 +3494,42 @@ abstract Expression implementSafe(RexToLixTranslator translator,
               gapInterval));
     }
   }
+
+  /** Implementor for bitwise  scalar function. */
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:3498]: extra space before `scalar `

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 6]
```diff
@@ -1170,6 +1180,25 @@ private static ByteString binaryOperator(
...
+   * Utility for bitwise function applied to byteString and long values.
+   *
+   * @param b0 The first byteString value operand of bitwise function.
+   * @param b1 The second long value operand of bitwise function.
+   * @param bitOp BitWise binary operator.
+   * @return ByteString after bitwise operation.
+   */
+  private static ByteString binaryOperator(
+      ByteString b0, long b1, BinaryOperator<Byte> bitOp) {
+    byte[] bytes0 = b0.getBytes();
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1193]: final?

**Comment by wangxlong** [comment]
@hsyuan Updated.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 7]
```diff
@@ -1170,6 +1180,25 @@ private static ByteString binaryOperator(
...
+   * @param b0 The first byteString value operand of bitwise function.
+   * @param b1 The second long value operand of bitwise function.
+   * @param bitOp BitWise binary operator.
+   * @return ByteString after bitwise operation.
+   */
+  private static ByteString binaryOperator(
+      ByteString b0, long b1, BinaryOperator<Byte> bitOp) {
+    final byte[] bytes0 = b0.getBytes();
+
+    final byte[] result = new byte[bytes0.length];
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1195]: Could you please remove `result` allocation and replace it with `bytes0`?

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Yes, Updated.

**Comment by wangxlong** [comment]
@hsyuan @vlsi  Hi, could you help me to review again, thanks.

**Comment by mihaibudiu** [comment]
What's the status of this PR? It looks like it's close to being ready for merging.

**Comment by mihaibudiu** [comment]
Superseded by https://github.com/apache/calcite/pull/3942

**Pull request closed without merging**